Add section about variance and positions (#6105)
dart-lang/site-www: src/content/resources/glossary.md

class C extends D {} // C is a subtype AND a subclass of D.
```

## Variance and variance positions

> A type parameter of a class (or a mixin, etc) is said to be _covariant_

Author: I think that's fine.

I plan to introduce a specific phrase for this in the language specification, perhaps 'type-introducing declaration'. The point is that a class declaration introduces a class, and it also introduces a type (and it introduces a family of classes and types if it's generic). A class is an entity that exists at run-time, and an entity which is capable of creating new objects. On top of that, it also introduces a namespace that contains static members. We don't actually have any (pure) type declarations, it's always declaring several other things plus a type that describes those other things. But I think it'll be OK to say 'type declaration' here.

when the type as a whole "co-varies" with the actual type argument, that
is: if we replace the type argument by a subtype then the type as a whole

> is also a subtype.

Author: Done.

For example, the type parameter of the class `List` is covariant because
list types co-vary with their type argument: `List<int>` is a subtype of
`List<Object>` because `int` is a subtype of `Object`.

In Dart, all type parameters of all class, mixin, mixin class, and enum
declarations are covariant.

> However, function types are different: A function type is covariant in the

Author: Indeed; done.

return type, but the opposite (known as _contravariant_) in its parameter
types.  For example, the type `int Function(int)` is a subtype of the type
`Object Function(int)`, but it is a supertype of `int Function(Object)`.

This makes sense if you consider their substitutability (see the section

> about subtypes): If you expect at compile time to work with a function of

Author: Of course, I think I was looking for the actual syntax used to write a link in this context, but forgot. I think it's correct now: `[substitutability](#subtype)`. Done.

type `int Function(int)` then it is OK to actually work with a function of
type `int Function(Object)` at run time: You will pass an `int` to it (and
that's fine, it accepts any `Object`), and it returns an `int` (which is

> exactly what you'd expect). Hence, `int Function(Object)` is a subtype of

Author: Used that, and did one more iteration on the text.

`int Function(int)`. Note that everything is turned upside-down when we

> consider a parameter type.

Reviewer: By "turned upside-down" do you mean "reversed" (as in the opposite) and what does "everything" here mean? The subtype relationship described in the preceding prose or the logic followed?

Author: The point is that if `X` occurs in a covariant position in `T` then `X` occurs in a contravariant position in `void Function(T)`, and if `Y` occurs in a contravariant position in `T` then `Y` occurs in a covariant position in `void Function(T)`. Finally, if `Z` occurs in an invariant position in `T` then it occurs in an invariant position in `void Function(T)`.

That's rather long and tedious, so I just said "everything" is reversed.

I don't know if it's a good idea to elaborate on this (to make this particular point explicit) or it's better to leave it as imprecise as it is currently. I suspect that the nested positions are much, much rarer than the basic ones (so people will write and/or encounter types like `void Function(T)`, but they are much less likely to encounter `void Function(void Function(T))`). If this is true then they won't ever need to use this "reverse everything" rule.

I expanded a bit on the text here.

When we consider a more complex type like `List<int Function(int)>` we need
to talk about the _positions_ in the type. This is basically just a matter
of turning one of the parts of the type into a placeholder (let's use `_`

> for that), and then consider what happens to the type when we put different

Author: Done.

types into that location in the type.

For example, consider `List<int Function(_)>` as a template for a type

> where we can put different types into the location where `_` occurs.

Reviewer: ```suggestion
where you can put different types in place of the placeholder (`_`).
```

Author: Done.

> This type is contravariant in that position: For example, 

Reviewer: Consider combining these two paragraphs, as the second one heavily relies on the first one. With the separation, it might not be clear what "This type" or "that position" refers to.

Reviewer: Also is the following example meant to explain _why_ its contravariant in that position or is it _because_ it's contravariant in that position?

My understanding is that it's an example of understanding _why_ that position is contravariant, but it's not clear to me from reading. Perhaps flipping the order of the because clauses and starting at the `Object` and `int` parameter type relation, might make that more clear.

Author: Combined the two paragraphs.

I changed the return type from `int` to `void` in the following examples, in order to be able to talk about `int` as a type that is unambiguously associated with a single position in the type (avoiding the need to say "the first `int`" vs. "the second `int`" and such).

> is the following example meant to explain _why_ its contravariant in that position

The text is intended to explain how 'contravariant position' is defined, by example. This is in itself an example, because it is intended to define 'contravariant position' as well as 'covariant position', based on a generalization which the reader would probably perform automatically, based on the previous explanations. I added a paragraph where 'covariant position' is defined briefly, but explicitly, in order to improve on this situation.

The text defines 'contravariant position' by explaining why we must conclude that the overall type `List<void Function(_)>` varies in the opposite direction of the variation at `_`, and then states that this by definition makes the position of `_` contravariant.

`List<int Function(Object)>` is a subtype of `List<int Function(int)>`
because `int Function(Object)` is a subtype of `int Function(int)` because
`int` is a subtype of `int` (the return type) and `Object` is a _supertype_
of `int` (the parameter type).

> This is the general case. In practice, it's often sufficient to know that

Reviewer: The general case of what? The position substitution exercise?

Author: I removed that sentence. I think the new text is clearer. ;-)

return type, but the opposite (known as _contravariant_) in its parameter
types.  For example, the type `int Function(int)` is a subtype of the type
`Object Function(int)`, but it is a supertype of `int Function(Object)`.

This makes sense if you consider their [substitutability](#subtype). If you
call a function with a static type of `int Function(int)`, that function
can actually be of type `int Function(Object)` at runtime. Based on the
static type, you expect to be able to pass an `int` to it. That will be
fine since the function actually accepts any `Object`, and this includes
every object of type `int`. Similarly, the returned result will be of type
`int`, which is also what you expect based on the static type.

Hence, `int Function(Object)` is a subtype of `int Function(int)`.

Note that everything is turned upside-down for parameter types. In
particular, this subtype relation among function types requires that the
_opposite_ subtype relation exists for the parameter type (for example,
`void Function(Object)` is a subtype of `void Function(int)` because `int`
is a subtype of `Object`).

With a more complex type like `List<void Function(int)>`, you have to

> consider the _positions_ in the type. This is basically just a matter of
> turning one of the parts of the type into a placeholder (let's use `_` for
> that), and then consider what happens to the type when different types are

Reviewer: Generally it's best to avoid phrasing like "basically just" as it don't add much meaning, and this process might not be basic for some readers. https://developers.google.com/style/word-list#simple

How about we be more prescriptive in how to do this:

```suggestion
consider the _positions_ in the type. To accomplish this,
turn one of the parts of the type into a placeholder,
and then consider what happens to the type when different types are
```

In this suggestion I also proposed dropping the "(let's use `_` for that)" to keep this sentence shorter and focused. The next paragraph can introduce that choice.

Author: Great point, thanks!

placed in that position.

For example, consider `List<void Function(_)>` as a template for a type
where you can put different types in place of the placeholder `_`. This
type is contravariant in the position where that placeholder occurs.

> We can illustrate this as follows. `List<void Function(Object)>` is a

Reviewer: With your adjustments, it's much more clear what this example is illustrating, thanks!

It might help further if we're extra specific about how we applied the suggested process in this example. Perhaps something like:

```
The following logic illustrates this by substituting `Object` and `int` for `_`.
```

This reference back might also help readers who jumped ahead know they should go back to the previous paragraph for more context.

Author: That works, thanks!
